examples/logisticregression/main.go

```diff
@@ -1,185 +1,288 @@
+	// binary cross entropy: -y * log(prob) - (1-y)*log(1-prob)
+	logProb := G.Must(G.Log(prob))
+	fstTerm := G.Must(G.HadamardProd(G.Must(G.Neg(y)), logProb))
+	oneMinusY := G.Must(G.Sub(one, y))
+	logOneMinusProb := G.Must(G.Log(G.Must(G.Sub(one, prob))))
+	sndTerm := G.Must(G.HadamardProd(oneMinusY, logOneMinusProb))
+
+	crossEntropy := G.Must(G.Sub(fstTerm, sndTerm))
+	G.WithName("crossEntropy")(crossEntropy)
+	loss := G.Must(G.Mean(crossEntropy))
+
+	// regularization term
+	weightSq := G.Must(G.Square(w))
+	sumSq := G.Must(G.Sum(weightSq))
+	l1reg := G.NewConstant(0.01, G.WithName("l1reg"))
+	regTerm := G.Must(G.Mul(l1reg, sumSq))
+
+	// cost we want to minimize
+	cost := G.Must(G.Add(loss, regTerm))
+	G.WithName("cost")(cost)
+
+	// calculate gradient by backpropagation https://en.wikipedia.org/wiki/Backpropagation
+	grads, err := G.Grad(cost, w, b)
+	handleError(err)
+	// "dcost/dw" == derivative of cost with respect to w
+	G.WithName("dcost/dw")(grads[0])
+	// "dcost/db" == derivative of cost with respect to b
+	G.WithName("dcost/db")(grads[1])
+
+	// create the nodes for calculating the gradient
+	learnRate := G.NewConstant(0.1) // be careful not to set a learnRate too high
+	gwlr := G.Must(G.Mul(learnRate, grads[0]))
+	wUpd := G.Must(G.Sub(w, gwlr))
+	gblr := G.Must(G.Mul(learnRate, grads[1]))
+	bUpd := G.Must(G.Sub(b, gblr))
+
+	// create the nodes that do the final operation of updating w and b
+	G.Set(w, wUpd)
+	G.Set(b, bUpd)
+	// write to the gographviz file for debugging https://github.com/awalterschulze/gographviz
+	ioutil.WriteFile("fullGraph.dot", []byte(g.ToDot()), 0644)
+
+	// Now that we have created all the notes, we can compile a program.
+	// We are essentially creating a list of instructions to get from our inputs {x, y}
+	// to our outputs {wUpd, bUpd}. Note that we only need to create the progam once.
+	prog, locMap, err := G.CompileFunction(g, G.Nodes{x, y}, G.Nodes{wUpd, bUpd})
+	handleError(err)
+	fmt.Printf("%v", prog) // print the instructions
+	// With our program, we initialize a new TapeMachine that will execute our program
+	machine := G.NewTapeMachine(g, G.WithPrecompiled(prog, locMap))
+
+	// we allocated the node w before, but we never set it's initial value
+	machine.Let(w, wT)
+	machine.Let(b, 0.0)
+	// ?? Why do we need to profile the CPU?
+	if *cpuprofile != "" {
+		f, err := os.Create(*cpuprofile)
+		if err != nil {
+			log.Fatal(err)
+		}
+		pprof.StartCPUProfile(f)
+		defer pprof.StopCPUProfile()
+	}
+
+	// Now that we have our graph, program, and machine, we can start training
+	start := time.Now()
+	for i := 0; i < trainIter; i++ {
+		// move the pointer back to the beginning of the prog. Reset() does not delete any values
+		machine.Reset()
+		// ?? why do we need to reinitialize the values the nodes {x,y}?
+		machine.Let(x, xT)
+		machine.Let(y, yT)
+		handleError(machine.RunAll())
+		// ?? Shouldn't the program have already done this? The last node has this Set() operation

```

Comment:
Actually the previous `G.Set()` were superfluous.  You can remove the previous ones.  They're superfluous because when the program is compiled the last nodes are `wUpd`, not the return value of `G.Set()` above

Reply:
removed